zerver/openapi/javascript_examples.py

```diff
@@ -0,0 +1,168 @@
+
+from typing import Dict, Any, Callable, Set, List
+
+import json
+import os
+import sys
+import subprocess
+
+from functools import wraps
+
+from zerver.openapi.openapi import validate_against_openapi_schema
+
+from zulip import Client
+
+TEST_FUNCTIONS = dict()  # type: Dict[str, Callable[..., None]]
+REGISTERED_TEST_FUNCTIONS = set()  # type: Set[str]
+CALLED_TEST_FUNCTIONS = set()  # type: Set[str]
+
+def openapi_test_function(endpoint: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """This decorator is used to register an openapi test function with
+    its endpoint. Example usage:
+
+    @openapi_test_function("/messages/render:post")
+    def ...
+    """
+    def wrapper(test_func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(test_func)
+        def _record_calls_wrapper(*args: Any, **kwargs: Any) -> Any:
+            CALLED_TEST_FUNCTIONS.add(test_func.__name__)
+            return test_func(*args, **kwargs)
+
+        REGISTERED_TEST_FUNCTIONS.add(test_func.__name__)
+        TEST_FUNCTIONS[endpoint] = _record_calls_wrapper
+
+        return _record_calls_wrapper
+    return wrapper
+
+def run_js_code(js_code: str) -> List[Dict[str, Any]]:
+    """Here, we run the JavaScript examples and return
+    a list of all the responses from the code example
+    as json."""
+
+    def get_json_data(response_list: List[str]) -> List[Dict[str, Any]]:
+
+        # Converts JavaScript objects to JSON.
+        js_code = """
+var responses = {response_list};
+responses.forEach(
+    (response) => console.log(JSON.stringify(eval("(" + response + ")")))
+);
+""".format(response_list=response_list)
+
+        output = subprocess.check_output(['node'],
+                                         input=js_code,
+                                         universal_newlines=True)
+        json_response_list = [json.loads(response) for response in output.splitlines()]
+
+        return json_response_list
+
+    # Response from 'check_output' here is a string with JavaScript objects
+    # which gets parsed to json responses by 'get_json_data'.
+    responses_str = subprocess.check_output(['node'], input=js_code, universal_newlines=True)
+    json_response_list = get_json_data(responses_str.splitlines())
```

Comment:
@timabbott We pass the response as a list of strings here. So those strings with JS objects get parsed using eval(). I think there might be some other way to do this. I'll try to make this work without eval().

Reply:
What if we extended the original `js_code` input with the `console.log` and `JSON.stringify()` statements, rather than doing a second invocation of `node`, and then just captured the JSON after that?

Reply:
@timabbott Initially, I did it that way, changing the `console.log` line to something like,
`client.messages.send(params).then((data)=>{
          console.log(JSON.stringify(data))
});`
But then the rendered code would include this.

Reply:
We definitely don't want that rendered in the docs.  But can't we append code to the thing that was rendered in the docs here, before running it?

Reply:
It might be more readable to document `async`/`await` syntax.

```js
const client = await zulip(config);
// Send a message
const params = {
    to: 'Denmark',
    type: 'stream',
    subject: 'Castle',
    content: 'I come not, friends, to steal away your hearts.',
};
console.log(JSON.stringify(await client.messages.send(params)));
```